sdk/dotnet/examples/bucket/TfGenStyleResources.cs

```diff
@@ -0,0 +1,61 @@
+// These are some hand authored resources in the style of what we think we'd generate via `tfgen`.  So we'll get
+// the shape right "by hand" and then work on the code-gen to stub everything else out:
+
+using Pulumi;
+using Pulumi.Rpc;
+
+namespace AWS.S3
+{
+    public class BucketObject : CustomResource
+    {
+        public BucketObject(string name, BucketObjectArgs args, ResourceOptions opts = null)
+            : base("aws:s3/bucketObject:BucketObject", name, args, opts)
+        {
+            OnConstructorCompleted();
+        }
+    }
+
+    public class BucketObjectArgs : ResourceArgs
+    {
+        public Input<string> Acl;
+        public Input<Id> Bucket;
+        public Input<string> ContentBase64;
+        public Input<string> ContentType;
+        public Input<string> Key;
+        public Input<AssetOrArchive> Source;

```

Comment:
How do we indicate the required properties? The program shouldn't compile until I set all required properties, e.g. (Azure example) I don't want this to go through because it has no resource group:
``` cs
var storageAccount = new Account("sa", new AccountArgs());
```

Reply:
> How do we indicate the required properties?

Comments would work.  Alternatively, explicitly using `Input<...>?` on the type.  We should definitely generate checks in the autogenerated code that the user is supplying all the required args.

Reply:
Comments are more of a last resort. Seeing red squiggles until you specify all required props is a life saver.

Reply:
> Seeing red squiggles until you specify all required props is a life saver.

We may have no recourse there except to use NRTs.  we can experiment to see waht works well there.

Reply:
I do agree this feels like a big loss in type-checking vs TypeScript. I’d say on the order of 50% of the overall type checking value we get comes from this validation. 

That said - it looks like the AWS .NET SDK behaves the same as the current implementation here and does not validate required properties statically.

Would definitely be interesting to understand whether Nullable Reference Types will allow us to smoothly improve on top of the existing API design.

Reply:
this is an open space of awfulness in .net currently (even with NRT).  It's a big ask that people have that they want to have mutable initializable objects taht must have values set right after construction.  There is no answer yet.

note: we could have constructors here with a split of required/optional args.  However then we lose the ability to do things like collection-initializers.

so, no good answers currently.

Reply:
Issues on csharplang tracking this:
https://github.com/dotnet/csharplang/issues/2109
https://github.com/dotnet/csharplang/issues/2328

Reply:
Pat and I are using NRT with actual runtime checks for this.  it's the best solution we have today.  ideally in teh future C# adds some support for the concept of types that must be initialized by the user after creation.